torchvision/prototype/datapoints/_dataset_wrapper.py

```diff
@@ -0,0 +1,250 @@
+# type: ignore
+
+from __future__ import annotations
+
+import contextlib
+
+import functools
+from collections import defaultdict
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+import PIL.Image
+import torch
+from torch.utils.data import Dataset
+from torchvision import datasets
+from torchvision.prototype import datapoints
+from torchvision.prototype.transforms import functional as F
+
+__all__ = ["wrap_dataset_for_transforms_v2"]
+
+_WRAPPERS = {}
+
+
+# TODO: naming!
+def wrap_dataset_for_transforms_v2(dataset: datasets.VisionDataset) -> _VisionDatasetDatapointWrapper:
+    wrapper = _WRAPPERS.get(type(dataset))
+    if wrapper is None:
+        raise TypeError
+    return _VisionDatasetDatapointWrapper(dataset, wrapper)
+
+
+class _VisionDatasetDatapointWrapper(Dataset):
+    def __init__(self, dataset: datasets.VisionDataset, wrapper) -> None:
+        self.vision_dataset = dataset
+        self.wrapper = wrapper
+
+        # We need to disable the transforms on the dataset here to be able to inject the wrapping before we apply the
+        # transforms
+        self.transform, dataset.transform = dataset.transform, None
+        self.target_transform, dataset.target_transform = dataset.target_transform, None
+        self.transforms, dataset.transforms = dataset.transforms, None
+
+    def __getattr__(self, item: str) -> Any:
+        with contextlib.suppress(AttributeError):
+            return object.__getattribute__(self, item)
+
+        return getattr(self.vision_dataset, item)
+
+    def __getitem__(self, idx: int) -> Any:
+        # This gets us the raw sample since we disabled the transforms for the underlying dataset in the constructor
+        # of this class
+        sample = self.vision_dataset[idx]
+
+        sample = self.wrapper(self.vision_dataset, sample)
+
+        # We don't need to care about `transform` and `target_transform` here since `VisionDataset` joins them into a
+        # `transforms` internally:
+        # https://github.com/pytorch/vision/blob/2d92728341bbd3dc1e0f1e86c6a436049bbb3403/torchvision/datasets/vision.py#L52-L54
+        if self.transforms is not None:
+            sample = self.transforms(*sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        return len(self.vision_dataset)
+
+
+@functools.lru_cache(maxsize=None)
```

Comment:
Is a LRU really needed here? It's not like we're going to create thousands of dataset instances, and the categories lists are fairly small anyway?

Reply:
The wrapper is called for every sample that is pulled out of the dataset. Since we need to perform some operation rather than jsut attribute access for some datasets, we should probably only do it once. Otherwise, we should at least measure if we take a performance hit here.